### ISSUE TITLE
Add NewReaderWithStart() to wal struct

### ISSUE THREAD

**Issue by 0x5487** [rosedblabs/wal#10]
Usecase:
we are planning to build a snapshot of a service with WAL project.  The service needs to stop all writing events and get the latest of chunk position in the wal instance.  When the service experiences an outage or downtime, the service can rebuild the state from snapshot file and replay the events since latest chunk position.  I was wondering if we could add a `LastChunkPosition` function to wal struct?

something like:
`func (wal *WAL) LastChunkPosition() (*ChunkPosition, error)`

**Comment by roseduan** [comment]
Sorry I didn't fully understand what you mean.

I think the last chunk position is the end of the file. Even you get it, when you are rebuild the service, you will iterate all data in wal? if so, the last chunk position means nothing.

**Comment by 0x5487** [comment]
Yes, you are correct. Instead of iterating through all the data in the WAL, the service can use a snapshot approach to restore its state more efficiently. In this approach, a snapshot file is created with the current state up to a specific point, such as event2. While the snapshot is being created, new events continue to be written to the WAL (event3, event4, event5).

In the event of an outage or downtime, the service can restore its state from the snapshot file and then replay the new events since event2. This allows the service to rebuild its state up to event5 without the need to iterate through the entire WAL. Using snapshots can significantly reduce the time required for state restoration and improve the overall efficiency of the service.

Therefore, we plan to get the last chunk position of the WAL, when a snapshot is being created. :)

**Comment by roseduan** [comment]
But now the WAL only support iterate all data through `wal.NewReader()`.

According to your description, the workflow in my mind is:
1. when your service started, write the first event and get a chunkPosition, like `(0, 100)`, and you will save the position for later use
2. accept more incoming events...
3. when your service restarted, you can call wal.NewReader() and pass the (0, 100), it will only iterate events after that

so in conclusion, i think we should add a start position in `wal.NewReader()` and it can iterate data after the given position.

like this:
`func (wal *WAL) NewReader(startPosition *ChunkPosition) *Reader {}`

Does these meet you needs?

**Comment by 0x5487** [comment]
Oops, you are correct.  I just realized `wal.NewReader` function couldn't pass `ChunkPostion` as parameter at the moment.

As I can see, the function name `NewReader` already exists in WAL instance.  The following are the ideas that come to mind for me at the moment.

option1:
`func (wal *WAL) NewReader(startPosition ...*ChunkPosition) *Reader {}`

option2:
`func (wal *WAL) NewReaderWithXXXX(startPosition *ChunkPosition) *Reader {}`

Which one is better?

**Comment by roseduan** [comment]
I think option2 is more reasonable, and I will implement it later.

**Comment by roseduan** [comment]
I have add a function in wal:
```
// NewReaderWithStart returns a new reader for the WAL,
// and the reader will only read the data from the segment file
// whose position is greater than or equal to the given position.
func (wal *WAL) NewReaderWithStart(startPos *ChunkPosition) (*Reader, error) {}
```

You can pass the start ChunkPosition to iterate all data after that.

You can recheck whether it meets your need, if so, I will release a new version for use.

**Comment by 0x5487** [comment]
Bro, it looks good to me.  Thank you.

**Comment by roseduan** [comment]
> Bro, it looks good to me. Thank you.

You can use v1.2.0 for the latest code.
https://github.com/rosedblabs/wal/releases/tag/v1.2.0
